### DISCUSSION TITLE
Can I use `buildctl` with buildx?

### DISCUSSION THREAD

**Discussion by richard-hajek** [docker/buildx#1860] [Q&A] [answered]
Hey,

I wanted to interact more with the buildx BuildKit ( Implementation? Process? ) and found `buildctl` and `buildkitd` tool. Are these compatibile with buildx?

**Comment by thaJeztah** [comment, marked as answer]
buildx is a BuildKit client (so roughly the equivalent to `buildctl`); `buildctl` is also a client, but more targeted for debugging purposes, and not as "polished" as buildx

**Reply by richard-hajek** [reply]
I see okay. Does `buildctl` and `buildx` operate over the same API/Library? I.e. can they be used interchangeably over the same data? Ie. if I build one image with buildctl, can buildx interact with it, and vice versa?

**Reply by thaJeztah** [reply]
Both are clients for the BuildKit grpc API, so (afaik) should be able to connect to the same BuildKit instance (but I must admit it's been a while since I used buildctl)

Is there a specific reason you want to use buildctl?

**Reply by richard-hajek** [reply]
Okay good! I wanted to know because I feel generally confused by the buildx vs buildctl vs buildkit vs docker relationship. I am also trying to do #1859 and wanted to use some debug commands from buildctl, because buildx doesn't seem to provide tools for that as far as I can tell.

**Reply by richard-hajek** [reply]
Specifically, access to build caches

**Reply by richard-hajek** [reply]
*Access to build caches beyond just pruning*

**Reply by richard-hajek** [reply]
Well apparently, neither does buildctl so I guess this solved itself
